modules/crypt.cpp

```diff
@@ -43,16 +42,151 @@
 #include <znc/Chan.h>
 #include <znc/User.h>
 #include <znc/IRCNetwork.h>
+#include <openssl/dh.h>
+#include <openssl/bn.h>
+#include <znc/SHA256.h>
 
 #define REQUIRESSL 1
 #define NICK_PREFIX_KEY "[nick-prefix]"
 
 class CCryptMod : public CModule {
+  private:
+    const char *prime1080="FBE1022E23D213E8ACFA9AE8B9DFADA3EA6B7AC7A7B7E95AB5EB2DF858921FEADE95E6AC7BE7DE6ADBAB8A783E7AF7A7FA6A2B7BEB1E72EAE2B72F9FA2BFB2A2EFBEFAC868BADB3E828FA8BADFADA3E4CC1BE7E8AFE85E9698A783EB68FA07A77AB6AD7BEB618ACF9CA2897EB28A6189EFA07AB99A8A7FA9AE299EFA7BA66DEAFEFBEFBF0B7D8B";
+    /* Generate our keys once and reuse, just like ssh keys */
+    DH *dh = NULL;
+    CString sPrivKey;
+    CString sPubKey;
+
+#if OPENSSL_VERSION_NUMBER < 0X10100000L
+    int DH_set0_pqg(DH *dh, BIGNUM *p, BIGNUM *q, BIGNUM *g) {
+        /* If the fields p and g in dh are NULL, the corresponding input
+         * parameters MUST be non-NULL.  q may remain NULL.
+         */
+        if (dh == NULL || (dh->p == NULL && p == NULL) || (dh->g == NULL && g == NULL))
+            return 0;
+
+        if (p != NULL) {
+            BN_free(dh->p);
+            dh->p = p;
+        }
+        if (g != NULL) {
+            BN_free(dh->g);
+            dh->g = g;
+        }
+        if (q != NULL) {
+            BN_free(dh->q);
+            dh->q = q;
+            dh->length = BN_num_bits(q);
+        }
+
+        return 1;
+    }
+
+    void DH_get0_key(const DH *dh, const BIGNUM **pub_key, const BIGNUM **priv_key) {
+        if (dh != NULL) {
+            if (pub_key != NULL)
+                *pub_key = dh->pub_key;
+            if (priv_key != NULL)
+                *priv_key = dh->priv_key;
+        }
+    }
+
+#endif
+
+    bool DH1080_gen() {
+        /* Generate our keys on first call */
+        if (!dh) {
+            int len;
+            unsigned char *buf;
+            const BIGNUM *bPrivKey = NULL;
+            const BIGNUM *bPubKey = NULL;
+            BIGNUM *bPrime = NULL;
+            BIGNUM *bGen = NULL;
+            dh = DH_new();
+
+            if (!BN_hex2bn(&bPrime, prime1080) || !BN_dec2bn(&bGen, "2") || !DH_set0_pqg(dh, bPrime, NULL, bGen) || !DH_generate_key(dh)) {
```

Comment:
Sorry, you are right. `DH_generate_key` doesn't actually sound as expensive as I previously thought.